L1Trigger/L1THGCal/src/be_algorithms/HGCalShowerShape.cc

```diff
@@ -9,536 +9,401 @@
-    //compute weighted eta RMS
-    float deltaphi2_sum=0;
-    for(int i=0;i<ntc;i++){
-        float deltaPhi=std::abs(phi.at(i)-phi_mean);
-        if (deltaPhi>TMath::Pi()) deltaPhi=2*TMath::Pi()-deltaPhi;
-        deltaphi2_sum+=energy.at(i)*pow(deltaPhi,2);
-    }        
-    float phi_RMS=0;
-    if (w_sum!=0) phi_RMS=deltaphi2_sum/w_sum;
-    float Spp=sqrt(phi_RMS);
-    return Spp;
-    
-}
 
 
 
-float HGCalShowerShape::sigmaZZ(const std::vector<float>& energy, const std::vector<float>& z) const {
+float HGCalShowerShape::sigmaEtaEta(const std::vector<float>& energy, const std::vector<float>& eta) const {
 
-    int ntc=energy.size();
-    //compute weighted eta mean
-    float z_sum=0;
-    float w_sum=0; //here weight is energy
-    for(int i=0;i<ntc;i++){
-        z_sum+=energy.at(i)*z.at(i);
-        w_sum+=energy.at(i);
-    }
-    float z_mean=0;
-    if (w_sum!=0) z_mean=z_sum/w_sum;
-    //compute weighted eta RMS
-    float deltaz2_sum=0;
-    for(int i=0;i<ntc;i++) deltaz2_sum+=energy.at(i)*pow((z.at(i)-z_mean),2);
-    float z_RMS=0;
-    if (w_sum!=0) z_RMS=deltaz2_sum/w_sum;
-    float Szz=sqrt(z_RMS);
-    return Szz;
+  int ntc=energy.size();
+  //compute weighted eta mean
+  float eta_sum=0;
+  float w_sum=0; //here weight is energy
+  for(int i=0;i<ntc;i++){
+    eta_sum+=energy.at(i)*eta.at(i);
+    w_sum+=energy.at(i);
+  }
+  float eta_mean=0;
+  if (w_sum!=0) eta_mean=eta_sum/w_sum;
+  //compute weighted eta RMS
+  float deltaeta2_sum=0;
+  for(int i=0;i<ntc;i++) deltaeta2_sum+=energy.at(i)*pow((eta.at(i)-eta_mean),2);
+  float eta_RMS=0;
+  if (w_sum!=0) eta_RMS=deltaeta2_sum/w_sum;
+  float See=sqrt(eta_RMS);
+  return See;
     
 }
 
 
-int HGCalShowerShape::showerLength(const l1t::HGCalMulticluster& c3d) const {
-
-    const edm::PtrVector<l1t::HGCalCluster>& clustersPtrs = c3d.constituents();
-
-    int firstLayer=999;
-    int lastLayer=-999;
-    int layer=999;
-
-    int nLayersEE=28;
-    int nLayersFH=12;
-   
-    for(const auto& clu : clustersPtrs){
-
-        if(clu->subdetId()==HGCEE) layer=clu->layer();//EE
-        if(clu->subdetId()==HGCHEF) layer=clu->layer()+nLayersEE;//FH
-        if(clu->subdetId()==HGCHEB) layer=clu->layer()+nLayersEE+nLayersFH;//FH
-
-        if(layer<firstLayer) firstLayer=layer;
-        if(layer>lastLayer) lastLayer=layer;
-
-    }
-
-    int nLayers=lastLayer-firstLayer+1;
-
-    return nLayers;
+float HGCalShowerShape::sigmaPhiPhi(const std::vector<float>& energy, const std::vector<float>& phi) const {
 
+  int ntc=energy.size();
+  //compute weighted phi mean
+  float phi_sum=0;
+  float w_sum=0; //here weight is energy
+  for(int i=0;i<ntc;i++){
+    if(phi.at(i)>0) phi_sum+=energy.at(i)*phi.at(i);
+    else phi_sum+=energy.at(i)*(phi.at(i)+2*TMath::Pi());
+    w_sum+=energy.at(i);
+  }
+  float phi_mean=0;
+  if (w_sum!=0) phi_mean=phi_sum/w_sum;
+  if(phi_mean>TMath::Pi()) phi_mean-=2*TMath::Pi();
+  //compute weighted eta RMS
```

Comment:
There is this above:
```cpp
    if(phi.at(i)>0) phi_sum+=energy.at(i)*phi.at(i);
    else phi_sum+=energy.at(i)*(phi.at(i)+2*TMath::Pi());
```
so `phi_mean` should be equal to `pi` (or `-pi` in case of rounding errors) in the example you mention

Reply:
Hmm, `-pi` and `pi` is not an issue but I think for instance -0.01 and 0.01 is.
With the implementation used so far (unchanged with this PR), after l.65, `phi_mean` will be equal to `pi`.
After l.66, depending on rounding, `phi_mean` will be equal to `pi` or `-pi` but in any case, not the expected value of 0.
I will prepare a fix for this.
Thanks for pointing this @amarini